libraries/core_libs/consensus/src/pbft/pbft_manager.cpp

```diff
@@ -1476,6 +1476,12 @@ std::optional<std::pair<blk_hash_t, uint64_t>> PbftManager::identifyLeaderBlock_
       continue;
     }
 
+    leader_candidates[getProposal(v)] = v;
+  }
+
+  for (auto const &leader : leader_candidates) {
+    const auto proposed_block_hash = leader.second->getBlockHash();
+
```

Comment:
hmm I did something very similar in [this PR](https://github.com/Taraxa-project/taraxa-node/pull/1997/files). These 3 checks are very cheap so they can be in the previous for loop:

- findPbftBlockInChain
- getUnverifiedPbftBlock
- checkPbftBlockValidation

So this second loop would look like this:

```
std::optional<std::pair<blk_hash_t, uint64_t>> PbftManager::identifyLeaderBlock_(uint64_t round, uint64_t period) {
  ...

  // Select first valid block from sorted leader candidates
  for (auto block_it = leader_candidates.begin(); block_it != leader_candidates.end(); block_it++) {
    if (compareBlocksAndRewardVotes_(block_it->second->getBlockHash())) {
      return {std::make_pair(block_it->second->getBlockHash(), block_it->second->getPeriod())};
    }

    LOG(log_er_) << "Incomplete or invalid proposed block " << block_it->second->getBlockHash() << ", period " << period
                 << ", round " << round;
  }

  LOG(log_er_) << "No valid leader candidate in period " << period << ", round " << round;
  return {}
}
```

Reply:
Even if they are cheap, why would we call them for each candidate if there is no need to do it?

Reply:
yes I believe they are, it is just checking if we have it in some structure and comparing previous pbft block hash